src/exo1 - Basic types/exo1.effect.solution.ts

```diff
@@ -0,0 +1,117 @@
+// Effect training Exercise 1
+// Basic types:
+// - Option
+// - Either
+// - Effect
+
+import { Option, Either, Effect, pipe } from "effect";
+import { sleep } from '../utils.effect';
+
+export const divide = (a: number, b: number): number => {
+  return a / b;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+//                                  OPTION                                   //
+///////////////////////////////////////////////////////////////////////////////
+
+// Write the safe version (meaning it handles the case where b is 0) of `divide` with signature:
+// safeDivide : (a: number, b: number) => Option<number>
+//
+// HINT: Option has two basic constructors:
+// - `O.some(value)`
+// - `O.none()`
+
+export const safeDivide = (a: number, b: number): Option.Option<number> => {
+  if (b === 0) {
+    return Option.none();
+  }
+
+  return Option.some(a / b);
+};
+
+// You probably wrote `safeDivide` using `if` statements, and it's perfectly valid!
+// There are ways to not use `if` statements.
+// Keep in mind that extracting small functions out of pipes and using `if` statements in them
+// is perfectly fine and is sometimes more readable than not using `if`.
+//
+// BONUS: Try now to re-write `safeDivide` without any `if`
+//
+// HINT: Have a look at `fromPredicate` constructor
+
+export const safeDivideBonus = (a: number, b: number): Option.Option<number> => {
+  // First create an Option from b, ensuring it's not zero
+  const nonZeroB = pipe(
+    Option.some(b),
+    Option.filter((n: number) => n !== 0)
+  );
+  
+  // Then map it to perform the division
+  return pipe(
+    nonZeroB,
+    Option.map((b) => a / b)
+  );
+};
+
+///////////////////////////////////////////////////////////////////////////////
+//                                  EITHER                                   //
+///////////////////////////////////////////////////////////////////////////////
+
+// Write the safe version of `divide` with signature:
+// safeDivideWithError : (a: number, b: number) => Either<DivideByZeroError, number>
+//
+// BONUS POINT: Implement `safeDivideWithError` in terms of `safeDivide`.
+//
+// HINT : Either has two basic constructors:
+// - `E.left(leftValue)`
+// - `E.right(rightValue)`
+// as well as "smarter" constructors like:
+// - `E.fromOption(() => leftValue)(option)`
```

Comment:
```suggestion
// - `Either.left(leftValue)`
// - `Either.right(rightValue)`
// as well as "smarter" constructors like:
// - `Either.fromOption(() => leftValue)(option)`
```

Reply:
Also I'm not sure when to mention them, but dual API in effect are frequent, so here you could use:
```ts
Either.fromOption(option, () => leftValue)
```